rviz_common/include/rviz_common/ros_integration/ros_node_abstraction.hpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2017, Open Source Robotics Foundation, Inc.
+ * Copyright (c) 2017, Bosch Software Innovations GmbH.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Willow Garage, Inc. nor the names of its
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef RVIZ_COMMON__ROS_INTEGRATION__ROS_NODE_ABSTRACTION_HPP_
+#define RVIZ_COMMON__ROS_INTEGRATION__ROS_NODE_ABSTRACTION_HPP_
+
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "./ros_node_abstraction_iface.hpp"
+
+namespace rviz_common
+{
+namespace ros_integration
+{
+
+// forward declaration so ros node storage headers can remain private
+class RosNodeStorageIface;
+
+class RosNodeAbstraction : public RosNodeAbstractionIface
+{
+public:
+  RosNodeAbstraction() = delete;
+
+  /// Creates a ros node with the given name
+  /**
+   * Internally a rclcpp::Node is created.
+   * If a rclcpp::Node with the given name already exists a node abstraction for this already
+   * existing node is created.
+   *
+   * \param node_name name of the node to create
+   */
+  explicit RosNodeAbstraction(const std::string & node_name);
```

Comment:
I sort of think it would be better to have the constructor private and only generate these instances via a `get_or_create_node_by_name()` (or some other name, just trying to be specific here) method of the `RosClientAbstrction` class.

Reply:
Introducing a static factory function does not allow mocking and thus kind of defeats the purpose of these changes. I agree that the current constructor solution is not ideal. I will try to expore using a separate factory.

Reply:
Ok, well don't spend too much time on it, if you come up with something that works better but still lets you mock it then great, otherwise we'll take it as-is.

The other thing to think about is how do you destroy nodes that have been created? In the case of ROS 2 (at least) you can make more than one node and you can remove them too. Maybe the `RosClientAbstraction` needs a destroy node method which is implicitly called on `shutdown()` if not already called for each node (typically just the one created during `init()`). Or the `RosNodeAbstraction` could be a scoped object (could be a `shared_ptr` with the node storage keeping only a `weak_ptr`). Dunno, just something to think about.